wpilibc/src/main/native/include/frc/controller/SimpleMotorFeedforward.h

```diff
@@ -50,12 +50,68 @@ class SimpleMotorFeedforward {
   constexpr units::volt_t Calculate(units::unit_t<Velocity> velocity,
                                     units::unit_t<Acceleration> acceleration =
                                         units::unit_t<Acceleration>(0)) const {
-    return m_kS * wpi::sgn(velocity) + m_kV * velocity + m_kA * acceleration;
+    return kS * wpi::sgn(velocity) + kV * velocity + kA * acceleration;
   }
 
- private:
-  units::volt_t m_kS{0};
-  units::unit_t<kv_unit> m_kV{0};
-  units::unit_t<ka_unit> m_kA{0};
+  // Rearranging the main equation from the calculate() method yields the
+  // formulas for the methods below:
+
+  /**
+   * Calculates the maximum achievable velocity given a maximum voltage supply
+   * and an acceleration.
+   *
+   * @param maxVoltage The maximum voltage that can be supplied to the motor.
+   * @param acceleration The acceleration of the motor.
+   * @return The maximum possible velocity at the given acceleration.
+   */
+  constexpr units::unit_t<Velocity> MaxAchievableVelocity(
+      units::volt_t maxVoltage, units::unit_t<Acceleration> acceleration) {
+    // Assume max velocity is positive
+    return (maxVoltage - kS - kA * acceleration) / kV;
```

Comment:
I'm not convinced that's the correct equation for max attainable velocity. Why does an acceleration value affect that? Is the interpretation of "attainable velocity" something different from what I'm expecting? Max attainable acceleration on the other hand makes sense due to back EMF represented by the kv term. :thinking:

Reply:
It's the instantaneous attainable velocity at the specified acceleration, per the javadoc comment.

Reply:
It makes sense because this is just the rearranged feedforward equation.

Reply:
But that's not correct. Your acceleration at that instant has nothing to do with the achievable velocity of the system. If you apply a given max voltage, your max achievable velocity is the steady-state velocity aka when the back EMF of the motor offsets the acceleration caused by the input voltage.

If you're wanting the max achievable velocity at the next timestep, the equation in the code here won't give that either. You'd have to integrate `a = -ks - kv/ka * v + 1/ka * maxV` forward one timestep.

Reply:
The "achievable velocity" is *given* that acceleration value - the same way the "achievable acceleration" is *given* a velocity value.  You're solving the voltage balance equation for a single variable, given fixed values for all the others.  What comes out is, indeed, physically meaningful - if not necessarily *useful*, in all cases.

This doesn't return "the maximum achievable velocity if you leave the system to respond to the input."  That's not what it purports to do.  We can add a "max steady-state velocity" method to do that, if you want.  It returns the the largest velocity you could conceivably achieve simultaneously with the given acceleration, given the voltage supply.

Of course, it is possible to specify accelerations that make this value silly and impractical - we can make the velocity as large as we like by passing in a sufficiently-large negative acceleration, but there's no non-contrived way to put the system into such a state.  So what?  It's still a state that obeys the system's governing equations.

If you want a concrete use case: say I know I want my system to be able to accelerate at 10 meters per second, and I know my battery cannot supply more than 12V.  What's the fastest my robot can go while still accelerating at the desired rate?

Reply:
Ok. You should put a variation of the last paragraph in the function docs so users know that's what the result means and know a use case. That wasn't at all obvious from the documentation alone.